### PR TITLE
Updated direct connection config APIs and docs

### PR DESCRIPTION
**Pull request by kushagraThapar**
* Updated direct connection config APIs and docs

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 1]
```diff
@@ -10,26 +10,25 @@
...
+    private int maxConnectionsPerEndpoint;
+    private int maxRequestsPerConnection;
 
     /**
      * Constructor.
      */
     public DirectConnectionConfig() {
         this.connectionTimeout = null;
-        this.idleChannelTimeout = DEFAULT_IDLE_CHANNEL_TIMEOUT;
+        this.idleConnectionTimeout = Duration.ZERO;
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:28]: what does 0 as idle connection timeout mean?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 2]
```diff
@@ -10,26 +10,25 @@
...
-    private int maxChannelsPerEndpoint;
-    private int maxRequestsPerChannel;
+    private int maxConnectionsPerEndpoint;
+    private int maxRequestsPerConnection;
 
     /**
      * Constructor.
      */
     public DirectConnectionConfig() {
         this.connectionTimeout = null;
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:27]:
```java
this.connectionTimeout = null;	
```
why is this null?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:28]: I have referred this in docs in `get/set` API.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 3]
```diff
@@ -98,49 +117,57 @@ public DirectConnectionConfig setIdleEndpointTimeout(Duration idleEndpointTimeou
...
     /**
-     * Gets the max requests per endpoint
+     * Gets the max requests per connection per endpoint
+     * Default value is 10
+     *
      * @return the max requests per endpoint
      */
-    public int getMaxRequestsPerChannel() {
-        return maxRequestsPerChannel;
+    public int getMaxRequestsPerConnection() {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:147]: if the user switches to DirectHttps some of these will not be relevant.

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:147]: Yups, `DirectHttps` is anyway not supported programatically. We also don't want to encourage it, but I get your point. 
May be we should call this config as `DirectTcpConnectionConfig` ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 2 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:27]: will get rid of the redundant initializer.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 4]
```diff
@@ -41,18 +39,20 @@ public static DirectConnectionConfig getDefaultConfig() {
         return new DirectConnectionConfig();
     }
 
-    //  TODO: (DANOBLE) - To update these docs.
-
     /**
-     * Gets the direct connection timeout
+     * Gets the connection timeout for direct client,
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:43]: Is 'direct client' mean Direct TCP connection? 
Client & connection overload confusing.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 5]
```diff
@@ -62,25 +62,38 @@ public DirectConnectionConfig setConnectionTimeout(Duration connectionTimeout) {
     }
 
     /**
-     * Gets the idle channel timeout
-     * @return idle channel timeout
+     * Gets the idle connection timeout for direct client
+     * Default value is {@link Duration#ZERO}
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:66]: Is it to keep the connections warm? 

How are broken connections identified? How will it impact the latencies/performance?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 6]
```diff
@@ -41,18 +39,20 @@ public static DirectConnectionConfig getDefaultConfig() {
         return new DirectConnectionConfig();
     }
 
-    //  TODO: (DANOBLE) - To update these docs.
-
     /**
-     * Gets the direct connection timeout
+     * Gets the connection timeout for direct client,
+     * represents timeout for establishing connections with an endpoint
+     *
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:45]: Guidance on when to tune to what values will help CX.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 7]
```diff
@@ -41,18 +39,20 @@ public static DirectConnectionConfig getDefaultConfig() {
...
      * @return direct connection timeout
      */
     public Duration getConnectionTimeout() {
         return connectionTimeout;
     }
 
     /**
-     *  Sets the direct connection timeout
+     *  Sets the connection timeout for direct client,
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:53]: Cover defaults details.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 8]
```diff
@@ -10,26 +10,24 @@
  */
 public final class DirectConnectionConfig {
     //  Constants
-    private static final Duration DEFAULT_IDLE_CHANNEL_TIMEOUT = Duration.ZERO;
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:13]: Include references to the Java perf best practices.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 9]
```diff
@@ -98,49 +116,57 @@ public DirectConnectionConfig setIdleEndpointTimeout(Duration idleEndpointTimeou
     }
 
     /**
-     * Gets the max channels per endpoint
-     * @return the max channels per endpoint
+     * Gets the max connections per endpoint
+     * Default value is 30
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:120]: Am not fully clear on endpointIdletimeout vs connecitonideltimeout. 

Lets please cover at depth the details. 
How about a conceptual MD file in Github which can be referenced?
We can extend it with patterns and guidance (some parts might eventually get promoted to docs). 

Also please refer the performance guide references across code docs extensively.

/cc: @David-Noble-at-work , @moderakh

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 3 above)
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:146]: Direct from CX will always be TCP. 
Https might be hacky workaround which is best avoided (Service team doesn't recommended it). 

Refering HTTPS in code docs if-any will lead a ambiguity/confuson. 

If Https is configured and some are not applicable, one options to fail at runtime which is okey.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 10]
```diff
@@ -41,18 +39,20 @@ public static DirectConnectionConfig getDefaultConfig() {
...
-    //  TODO: (DANOBLE) - To update these docs.
-
     /**
-     * Gets the direct connection timeout
+     * Gets the connection timeout for direct client,
+     * represents timeout for establishing connections with an endpoint
+     *
      * @return direct connection timeout
      */
     public Duration getConnectionTimeout() {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:48]: How about stating how these manifest for netty underlying for advanced users?

Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:48]: connection vs socket vs others: Is there guidance or patterns from azure SDK's?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 4 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:43]: direct client means direct tcp client itself.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 6 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:45]: Guidance on these can vary based on customer, case to case. Referring material in perf guide will help and as you mentioned we can refer the perf guide in this docs.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 7 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:53]: Will do.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 8 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:13]: Cool, will do.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 9 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:120]: endpoint is a pool of connections per replica. I will improve the documentation around this. 
We initially had a MD file explaining Rntbd architecture. I will work with @David-Noble-at-work  to introduce it back if we can.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 10 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:48]: So far we have seen connection - as that is the easiest concept for an end user. 
Http client from azure-core -> https://github.com/Azure/azure-sdk-for-java/blob/fc2612029135148ad427083b8a24b672e634a1f6/sdk/core/azure-core-http-jdk-httpclient/src/main/java/com/azure/core/http/jdk/httpclient/JdkAsyncHttpClientBuilder.java#L73
It uses connectionTimeout as well.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 5 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:66]: Yes, to keep connections warm. When the connections are closed, rntbd logs it. 
We have seen a connection establishment takes anywhere between 200ms - 300ms .

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 10 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:48]: Sure, will attach respective Netty links here.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 11]
```diff
@@ -3,33 +3,35 @@
...
+ * For performance tips on how to optimize Direct connection configuration,
+ * refer to performance tips guide: https://docs.microsoft.com/en-us/azure/cosmos-db/performance-tips-java-sdk-v4-sql?tabs=api-async
  */
 public final class DirectConnectionConfig {
     //  Constants
-    private static final Duration DEFAULT_IDLE_CHANNEL_TIMEOUT = Duration.ZERO;
     private static final Duration DEFAULT_IDLE_ENDPOINT_TIMEOUT = Duration.ofSeconds(70L);
-    private static final int DEFAULT_MAX_CHANNELS_PER_ENDPOINT = 30;
+    private static final int DEFAULT_MAX_CONNECTIONS_PER_ENDPOINT = 30;
     private static final int DEFAULT_MAX_REQUESTS_PER_ENDPOINT = 10;
```
Comment (David-Noble-at-work) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:19]: DEFAULT_MAX_REQUESTS_PER_ENDPOINT should be DEFAULT_MAX_REQUESTS_PER_CONNECTION

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 12]
```diff
@@ -41,18 +43,28 @@ public static DirectConnectionConfig getDefaultConfig() {
...
+     *
      * @return direct connection timeout
      */
     public Duration getConnectionTimeout() {
         return connectionTimeout;
     }
 
     /**
-     *  Sets the direct connection timeout
+     * Sets the connection timeout for direct client,
```
Comment (David-Noble-at-work) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:61]: It is the idleConnectionTimeout that defaults to zero. The connectTimeout defaults to the value of requestTimeout: 60 seconds prior to this PR.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 11 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:19]: Good catch, thanks.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 12 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:61]: Cool, will add default value here - and will add the constant too..

**Pull request merged**